streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -648,71 +794,142 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
+            }
 
-                final Map<TaskId, Set<TopicPartition>> standbyTasks = new HashMap<>();
-                for (final TaskId taskId : clientMetadata.state.prevStandbyTasks()) {
-                    standbyTasks.put(taskId, partitionsForTask.get(taskId));
+            final Iterator<String> consumerIt = unfilledConsumers.iterator();
+            while (consumerIt.hasNext()) {
+                final String consumer = consumerIt.next();
+                final List<TaskId> consumerAssignment = assignments.get(consumer);
+                final TaskId task = newTasks.poll();
+                if (task == null) {
+                    break;
                 }
 
-                assignment.put(consumerId, new Assignment(
-                    assignedPartitions,
-                    new AssignmentInfo(
-                        minUserMetadataVersion,
-                        activeTasks,
-                        standbyTasks,
-                        partitionsByHostState,
-                        0)
-                        .encode()
-                ));
+                consumerAssignment.add(task);
+                if (consumerAssignment.size() == maxTasksPerClient) {
+                    consumerIt.remove();
+                }
             }
         }
 
-        // add empty assignment for "future version" clients (ie, empty version probing response)
-        for (final String consumerId : futureConsumers) {
-            assignment.put(consumerId, new Assignment(
-                Collections.emptyList(),
-                new AssignmentInfo().encode()
-            ));
-        }
+        return assignments;
+    }
 
-        return assignment;
+    /**
+     * Get the previous consumer for the partitions of a task
+     *
+     * @param taskPartitions the TopicPartitions for a single given task
+     * @param clientOwnedPartitions the partitions owned by all consumers in a client
+     * @param allOwnedPartitions all partitions claimed as owned by any consumer in any client
+     * @return the previous consumer that owned these partitions
+     *         empty string if it is a new task, or its owner is no longer in the group
+     *         null if its previous owner is from a different client or the task's partitions had different owners
+     */
+    String previousConsumerOfTaskPartitions(final Set<TopicPartition> taskPartitions,
+                                           final Map<TopicPartition, String> clientOwnedPartitions,
+                                           final Set<TopicPartition> allOwnedPartitions) {
+        String firstConsumer = "";
+        for (final TopicPartition tp : taskPartitions) {
+            // initialize the first consumer if this is the first iteration
+            if (firstConsumer != null && firstConsumer.equals("")) {
+                firstConsumer = clientOwnedPartitions.get(tp);
+            }
+
+            final String thisConsumer = clientOwnedPartitions.get(tp);
```

Comment:
Related to my other comment: since now `allOwnedPartitions` is just a union of `clientOwnedPartitions` I think it is not needed anymore inside the `assign` function.

Reply:
I believe we still need `allOwnedPartitions` here because we need to distinguish between partitions which are not owned _within_ this client vs partitions which are not owned by _anyone_ in any client. ie just because no one in the client claims a partition as owned doesn't mean it is safe to give that partition/task away, since it might still be owned by someone in the group and need to be revoked

Reply:
My read was that the keyset of `Map<TopicPartition, String> clientOwnedPartitions` and `Set<TopicPartition> allOwnedPartitions` are the same, but now I realized the former is only for a single client (as with my other comment above) actually.. Now this logic makes sense.

Reply:
It's hard to give good self-explanatory names to _all_ your variables 😉 Let me know if you think of any names that better explain what `allOwnedPartitions` is supposed to be